# Problem Statement
A large company named XYZ, employs, at any given point of time, around 4000 employees. However, every year, around 15% of its employees leave the company and need to be replaced with the talent pool available in the job market. The management believes that this level of attrition (employees leaving, either on their own or because they got fired) is bad for the company, because of the following reasons -

The former employees’ projects get delayed, which makes it difficult to meet timelines, resulting in a reputation loss among consumers and partners
A sizeable department has to be maintained, for the purposes of recruiting new talent
More often than not, the new employees have to be trained for the job and/or given time to acclimatise themselves to the company
Hence, the management has contracted an HR analytics firm to understand what factors they should focus on, in order to curb attrition. In other words, they want to know what changes they should make to their workplace, in order to get most of their employees to stay. Also, they want to know which of these variables is most important and needs to be addressed right away.

Since you are one of the star analysts at the firm, this project has been given to you.

Goal of the case study You are required to model the probability of attrition. The results thus obtained will be used by the management to understand what changes they should make to their workplace, in order to get most of their employees to stay.

Columns
EmployeeIDEmployee number/id
EnvironmentSatisfactionWork Environment Satisfaction Level
JobSatisfactionJob Involvement Level Job Involvement Level Job Involvement Level
WorkLifeBalanceWork life balance level

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('general_data.csv')
df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

In [5]:
df.describe()

Age  DistanceFromHome    Education  EmployeeCount   EmployeeID  \
count  4410.000000       4410.000000  4410.000000         4410.0  4410.000000   
mean     36.923810          9.192517     2.912925            1.0  2205.500000   
std       9.133301          8.105026     1.023933            0.0  1273.201673   
min      18.000000          1.000000     1.000000            1.0     1.000000   
25%      30.000000          2.000000     2.000000            1.0  1103.250000   
50%      36.000000          7.000000     3.000000            1.0  2205.500000   
75%      43.000000         14.000000     4.000000            1.0  3307.750000   
max      60.000000         29.000000     5.000000            1.0  4410.000000   

          JobLevel  MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  \
count  4410.000000    4410.000000         4391.000000        4410.000000   
mean      2.063946   65029.312925            2.694830          15.209524   
std       1.106689   47068.888559            2.498887           3.659108   
min       1.000000   10090.000000            0.000000          11.000000   
25%       1.000000   29110.000000            1.000000          12.000000   
50%       2.000000   49190.000000            2.000000          14.000000   
75%       3.000000   83800.000000            4.000000          18.000000   
max       5.000000  199990.000000            9.000000          25.000000   

       StandardHours  StockOptionLevel  TotalWorkingYears  \
count         4410.0       4410.000000        4401.000000   
mean             8.0          0.793878          11.279936   
std              0.0          0.851883           7.782222   
min              8.0          0.000000           0.000000   
25%              8.0          0.000000           6.000000   
50%              8.0          1.000000          10.000000   
75%              8.0          1.000000          15.000000   
max              8.0          3.000000          40.000000   

       TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
count            4410.000000     4410.000000              4410.000000   
mean                2.799320        7.008163                 2.187755   
std                 1.288978        6.125135                 3.221699   
min                 0.000000        0.000000                 0.000000   
25%                 2.000000        3.000000                 0.000000   
50%                 3.000000        5.000000                 1.000000   
75%                 3.000000        9.000000                 3.000000   
max                 6.000000       40.000000                15.000000   

       YearsWithCurrManager  
count           4410.000000  
mean               4.123129  
std                3.567327  
min                0.000000  
25%                2.000000  
50%                3.000000  
75%                7.000000  
max               17.000000

In [21]:
df['Attrition'].isnull().sum()

0

In [23]:
df['Attrition'].value_counts()

No     3699
Yes     711
Name: Attrition, dtype: int64

In [28]:
df[df['Attrition'] == 'Yes'].describe()

Age  DistanceFromHome   Education  EmployeeCount   EmployeeID  \
count  711.000000        711.000000  711.000000          711.0   711.000000   
mean    33.607595          9.012658    2.877637            1.0  2191.767932   
std      9.675693          7.772368    1.014233            0.0  1270.136654   
min     18.000000          1.000000    1.000000            1.0     2.000000   
25%     28.000000          2.000000    2.000000            1.0  1072.000000   
50%     32.000000          7.000000    3.000000            1.0  2165.000000   
75%     39.000000         15.000000    4.000000            1.0  3341.000000   
max     58.000000         29.000000    5.000000            1.0  4403.000000   

         JobLevel  MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  \
count  711.000000     711.000000          707.000000         711.000000   
mean     2.037975   61682.616034            2.936351          15.481013   
std      1.057485   44792.067695            2.678774           3.775289   
min      1.000000   10090.000000            0.000000          11.000000   
25%      1.000000   28440.000000            1.000000          12.000000   
50%      2.000000   49080.000000            1.000000          14.000000   
75%      2.000000   71040.000000            5.000000          18.000000   
max      5.000000  198590.000000            9.000000          25.000000   

       StandardHours  StockOptionLevel  TotalWorkingYears  \
count          711.0        711.000000         709.000000   
mean             8.0          0.780591           8.255289   
std              0.0          0.858899           7.164018   
min              8.0          0.000000           0.000000   
25%              8.0          0.000000           3.000000   
50%              8.0          1.000000           7.000000   
75%              8.0          1.000000          10.000000   
max              8.0          3.000000          40.000000   

       TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
count             711.000000      711.000000               711.000000   
mean                2.654008        5.130802                 1.945148   
std                 1.154834        5.941598                 3.148633   
min                 0.000000        0.000000                 0.000000   
25%                 2.000000        1.000000                 0.000000   
50%                 3.000000        3.000000                 1.000000   
75%                 3.000000        7.000000                 2.000000   
max                 6.000000       40.000000                15.000000   

       YearsWithCurrManager  
count            711.000000  
mean               2.852321  
std                3.138918  
min                0.000000  
25%                0.000000  
50%                2.000000  
75%                5.000000  
max               14.000000

In [31]:
df[df['Attrition'] == 'Yes']['Department'].value_counts()

Research & Development    453
Sales                     201
Human Resources            57
Name: Department, dtype: int64

Reasons could be
1. On an average people are working here for 8 years so they might be looking for a new job.
2. On an average people havent been promoted in 2 years.
3. Most people who left or laid off are in the department of Reseach and Development.
4. On an average people are with the same manager for over 2.5 years.